src/ClientData/include/ClientData/ModuleManager.h

```diff
@@ -19,29 +19,47 @@
 #include <vector>
 
 #include "ClientData/ModuleData.h"
-#include "ClientData/ProcessData.h"
+#include "ClientData/ModuleIdentifier.h"
+#include "ClientData/ModuleIdentifierProvider.h"
+#include "ClientData/ModuleInMemory.h"
 #include "ClientProtos/capture_data.pb.h"
 #include "GrpcProtos/module.pb.h"
-#include "SymbolProvider/ModuleIdentifier.h"
 #include "absl/container/node_hash_map.h"
 #include "absl/synchronization/mutex.h"
 
 namespace orbit_client_data {
 
 class ModuleManager final {
  public:
-  explicit ModuleManager() = default;
+  explicit ModuleManager(ModuleIdentifierProvider* module_identifier_provider)
+      : module_identifier_provider_{module_identifier_provider} {};
 
   [[nodiscard]] const ModuleData* GetModuleByModuleInMemoryAndAbsoluteAddress(
       const ModuleInMemory& module_in_memory, uint64_t absolute_address) const;
 
   [[nodiscard]] ModuleData* GetMutableModuleByModuleInMemoryAndAbsoluteAddress(
       const ModuleInMemory& module_in_memory, uint64_t absolute_address);
 
+  [[nodiscard]] const ModuleData* GetModuleByModulePathAndBuildId(std::string_view module_path,
+                                                                  std::string_view build_id) const {
+    std::optional<orbit_client_data::ModuleIdentifier> module_id =
+        module_identifier_provider_->GetModuleIdentifier(module_path, build_id);
+    if (!module_id.has_value()) return nullptr;
+    return GetModuleByModuleIdentifier(module_id.value());
+  }
+
+  [[nodiscard]] ModuleData* GetMutableModuleByModulePathAndBuildId(std::string_view module_path,
```

Comment:
I'm confused with the thread safety of these two methods. `module_identifier_provider_` has its own mutex internally, however `mutex_` of this class is taken before accessing `module_identifier_provider_` in `AddOrUpdateModules` and in `AddOrUpdateNotLoadedModules`, but not here. This feels problematic.

Reply:
Why do you think this is a problem?
The `mutex_` of the class exists to protect the two maps. Those will be protected in the `GetModuleByIdentifier` call.

I mean what could happen is the following: We add a bunch of modules in `AddOrUpdateModules`. During that time the mutex of this class is locked for the complete execution. This method may add multiple modules to the identifier provider. Each addition is locked there separately.
Now, there are concurrent calls to `GetModuleByModulePathAndBuildId`. Here the identifier provider locks its mutex. Assuming we query one of the "new" modules, the id might either already exist, or not. If it exists, then the `GetModuleByModuleIdentifier` will need to wait until the `AddOrUpdateModules` is done, as it will take a lock. So at that point, the module data will exist for that module identifier.

I think this is fine. Do you see something particularly problematic here?

Reply:
I have not considered every possible scenario, but I don't think I should. Maybe the code works now, but it's confusing that sometimes `module_identifier_provider_` is accessed under `mutex_` and sometimes it isn't: it feels like every change to this class would require you to think again why this is fine. So what is the downside of consistently locking everything with `mutex_`, which is how this classed worked before?

Reply:
I can see that. The issue is that the mutex is not recursive/reentrant. So we would need to have an `GetModuleByIdentifierInternal` that does not take the lock. Not sure if this is better.

Reply:
Why do you see that as an issue?

Reply:
I would argue that it is also not intuitive. But I am fine with it.

Reply:
It's extracting common code between `GetModuleByModulePathAndBuildId` and `GetModuleByModuleIdentifier` into a separate method. The new method can even be documented with `EXCLUSIVE_LOCKS_REQUIRED`. Why is it not intuitive?

Reply:
Yeah, I think you are right. Especially with the annotation its probably fine. Thanks!

Reply:
Done.